Block sparse qmm

angeloskath opened the pull request:
The `block_sparse_mm` equivalent for quantized matmuls. A big diff but mostly boilerplate code and refactoring.

jagrit06 commented on mlx/ops.cpp:
```diff
@@ -3430,6 +3469,51 @@ array dequantize(
+  out_shape.push_back(w_outer_dims);
+
+  // and output type
+  auto out_type = result_type(x, scales, biases);
```
Do we need to add a check or astype here to make sure all 3 are the same dtype ?

angeloskath replied:
Good catch I forgot to use it below to cast `x`, `scales` and `biases` to out_type.

jagrit06 commented on mlx/primitives.cpp:
```diff
@@ -2375,6 +2375,76 @@ bool QuantizedMatmul::is_equivalent(const Primitive& other) const {
+    else {
+      throw std::runtime_error(
+          "BlockSparseQMM::vjp no gradient wrt the quantized matrix yet.");
+    }
```
Maybe we can add a separate error message if gradients w.r.t to the indices are requested

jagrit06 commented on mlx/primitives.cpp:
```diff
@@ -2375,6 +2375,76 @@ bool QuantizedMatmul::is_equivalent(const Primitive& other) const {
+
+bool BlockSparseQMM::is_equivalent(const Primitive& other) const {
+  const BlockSparseQMM& qm_other = static_cast<const BlockSparseQMM&>(other);
+  return group_size_ == qm_other.group_size_ && bits_ == qm_other.bits_;
```
Should we also check `transpose_` here ?

jagrit06 reviewed (approved):
Looks great! Just a few small comments here and there

angeloskath commented:
@awni and @jagrit06 I made two changes that are not directly related to the `block_sparse_qmm` but general quantization. So I 'd appreciate a review. Also if you think that they should be on a different PR (or maybe just the 2nd change) let me know and I can merge this before these commits and make another PR for this.

The changes are the following:
1. `quantize` and `dequantize` now support arbitrary shapes instead of only 2D matrices
2. I added a `to_quantized` method to `Linear` and `Embedding` and changed `nn.quantized` to check for modules with the `to_quantized` method implemented. This allows arbitrary modules to support quantization using `nn.quantize` without MLX knowing anything about them.

awni reviewed (commented):
LGTM 🚀
